torch/_refs/linalg/__init__.py

```diff
@@ -0,0 +1,103 @@
+import torch
+
+import torch._prims as prims
+import torch._refs as refs
+from torch._prims.wrappers import out_wrapper
+
+from torch._prims.utils import (
+    DimsType,
+    TensorLikeType,
+    NumberType,
+    corresponding_real_dtype,
+    is_float_dtype,
+    get_computation_dtype,
+    reduction_dtypes,
+    REDUCTION_OUTPUT_TYPE_KIND,
+)
+import torch._refs.linalg as linalg
+import torch._refs.linalg.utils
+
+from typing import Optional, List
+from functools import partial
+
+__all__ = [
+    "vector_norm",
+]
+
+
+@out_wrapper
+def vector_norm(
+    x: TensorLikeType,
+    ord: float = 2.0,
+    dim: Optional[DimsType] = None,
+    keepdim: bool = False,
+    *,
+    dtype: Optional[torch.dtype] = None,
+):
+    # Checks
+    linalg.utils.check_fp_or_complex(x, "linalg.vector_norm", half=True)
+
+    if isinstance(dim, int):
+        dim = [dim]  # type: ignore[assignment]
+    elif not isinstance(dim, List) and dim is not None:
+        # refs.sum just accepts List rather than DimType
+        dim = list(dim)  # type: ignore[assignment]
+
+    # TODO These things are TORCH_CHECKS, replace them with check(a,b)
+    # once https://github.com/pytorch/pytorch/pull/78014 is merged
+    if x.numel() == 0 and (ord < 0.0 or ord == float("inf")):
+        if dim is None or len(dim) == 0:
+            raise RuntimeError(
+                "linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
+                "because the operation does not have an identity"
+            )
+        else:
+            for d in dim:
+                if x.size(d) == 0:
+                    raise RuntimeError(
+                        f"linalg.vector_norm cannot compute the {ord} norm on the "
+                        "dimension {d} because this dimension is empty and the "
+                        "operation does not have an identity"
+                    )
+    linalg.utils.check_norm_dtype(dtype, x.dtype, "linalg.vector_norm")
+
+    computation_dtype, result_dtype = reduction_dtypes(
+        x, REDUCTION_OUTPUT_TYPE_KIND.COMPLEX_TO_FLOAT, dtype
+    )
+
+    to_result_dtype = partial(prims.convert_element_type, dtype=result_dtype)
+
+    # Implementation
+    if ord == 0.0:
+        return refs.sum(refs.ne(x, 0.0), dim=dim, keepdim=keepdim, dtype=result_dtype)
+    elif ord == float("inf"):
+        return to_result_dtype(refs.amax(prims.abs(x), dim=dim, keepdim=keepdim))
+    elif ord == float("-inf"):
+        return to_result_dtype(refs.amin(prims.abs(x), dim=dim, keepdim=keepdim))
+    else:
+        # From here on the computation dtype is important as the reduction is non-trivial
+        x = prims.convert_element_type(x, computation_dtype)
+        reduce_sum = partial(refs.sum, dim=dim, keepdim=keepdim)
+
+        def fast_pow(x, p):
+            if p == 1.0:
+                return x
+            elif p == 2.0:
+                return prims.mul(x, x)
```

Comment:
Is this an optimisation we want, or is `pow(x, 2.0)` better? Same for `sqrt` below really

Reply:
I prefer the explicit multiplication because otherwise I think we're just making more work for ourselves to convert the power to a multiplication later

Reply:
I can see getting rid of square, but are pow(a, 1/2) and pow(a, 1/3) valid implementations for sqrt and cbrt? What we aren't doing at this time is requiring trace executors make numerical accuracy fixes, in part because we also want to run accurately in eager mode. If every system has to remap pow(a, 1/2) into sqrt, anyway, then aren't we just hiding the fact that it's prim?

Reply:
I would be against using `pow` because it requires unusual smarts from the backend. I don't care about `square` or `x*x`.